drivers/gpio/gpio_rcar.c

```diff
@@ -0,0 +1,326 @@
+/*
+ * Copyright (c) 2020 IoT.bzh
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT renesas_rcar_gpio
+
+#include <errno.h>
+
+#include <kernel.h>
+#include <device.h>
+#include <init.h>
+#include <drivers/gpio.h>
+#include <drivers/clock_control.h>
+#include <drivers/clock_control/rcar_clock_control.h>
+
+#include "gpio_utils.h"
+
+typedef void (*init_func_t)(const struct device *dev);
+
+struct gpio_rcar_cfg {
+	struct gpio_driver_config common;
+	uint32_t reg_addr;
+	int reg_size;
+	init_func_t init_func;
+	const struct device *clock_dev;
+	struct rcar_cpg_clk mod_clk;
+};
+
+struct gpio_rcar_data {
+	struct gpio_driver_data common;
+	sys_slist_t cb;
+};
+
+#define IOINTSEL 0x00   /* General IO/Interrupt Switching Register */
+#define INOUTSEL 0x04   /* General Input/Output Switching Register */
+#define OUTDT    0x08   /* General Output Register */
+#define INDT     0x0c   /* General Input Register */
+#define INTDT    0x10   /* Interrupt Display Register */
+#define INTCLR   0x14   /* Interrupt Clear Register */
+#define INTMSK   0x18   /* Interrupt Mask Register */
+#define MSKCLR   0x1c   /* Interrupt Mask Clear Register */
+#define POSNEG   0x20   /* Positive/Negative Logic Select Register */
+#define EDGLEVEL 0x24   /* Edge/level Select Register */
+#define FILONOFF 0x28   /* Chattering Prevention On/Off Register */
+#define OUTDTSEL 0x40   /* Output Data Select Register */
+#define BOTHEDGE 0x4c   /* One Edge/Both Edge Select Register */
+
+/* Helper macros for GPIO */
+#define DEV_GPIO_CFG(dev) \
+	((const struct gpio_rcar_cfg *)(dev)->config)
+#define DEV_GPIO_DATA(dev) \
+	((struct gpio_rcar_data *)(dev)->data)
+
+static inline uint32_t gpio_rcar_read(const struct gpio_rcar_cfg *config,
+				      uint32_t offs)
+{
+	return sys_read32(config->reg_addr + offs);
+}
+
+static inline void gpio_rcar_write(const struct gpio_rcar_cfg *config,
+				   uint32_t offs, uint32_t value)
+{
+	sys_write32(value, config->reg_addr + offs);
+}
+
+static void gpio_rcar_modify_bit(const struct gpio_rcar_cfg *config,
+				 uint32_t offs, int bit, bool value)
+{
+	uint32_t tmp = gpio_rcar_read(config, offs);
+
+	if (value) {
+		tmp |= BIT(bit);
+	} else {
+		tmp &= ~BIT(bit);
+	}
+
+	gpio_rcar_write(config, offs, tmp);
+}
+
+static void gpio_rcar_port_isr(void *arg)
+{
+	const struct device *dev = (const struct device *)arg;
+	const struct gpio_rcar_cfg *config = DEV_GPIO_CFG(dev);
+	struct gpio_rcar_data *data = DEV_GPIO_DATA(dev);
+	uint32_t pending, fsb, mask;
+
+	pending = gpio_rcar_read(config, INTDT);
+	mask = gpio_rcar_read(config, INTMSK);
+
+	while ((pending = gpio_rcar_read(config, INTDT) &
+			  gpio_rcar_read(config, INTMSK))) {
+		fsb = find_lsb_set(pending) - 1;
+		gpio_rcar_write(config, INTCLR, BIT(fsb));
```

Comment:
Should it `INTCLR` after all the callbacks are processed in order to prevent nested entry into the `gpio_rcar_port_isr` and the callbacks?

Reply:
Indeed that's better, I saw this scheme on other driver but interrupt handling may be different.

Reply:
fixed